pkg/reconciler/apis/permissionclaimlabel/permissionclaimlabel_controller.go

```diff
@@ -0,0 +1,226 @@
+)
+
+const (
+	controllerName = "kcp-permissionclaimlabel"
+)
+
+// NewController returns a new controller for handling permission claims for an APIBinding.
+// it will own the ObservedAcceptedPermissionClaims and will own the accepted permision claim condition.
+func NewController(
+	kcpClusterClient kcpclient.Interface,
+	dynamicClusterClient dynamic.Interface,
+	dynamicDiscoverySharedInformerFactory *informer.DynamicDiscoverySharedInformerFactory,
+	apiBindingInformer apisinformers.APIBindingInformer,
+) (*controller, error) {
+	queue := workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), controllerName)
+
+	c := &controller{
+		queue:                queue,
+		kcpClusterClient:     kcpClusterClient,
+		dynamicClusterClient: dynamicClusterClient,
+		ddsif:                dynamicDiscoverySharedInformerFactory,
+
+		apiBindingsLister: apiBindingInformer.Lister(),
+		listAPIBindings: func(clusterName logicalcluster.Name) ([]*apisv1alpha1.APIBinding, error) {
+			var ret []*apisv1alpha1.APIBinding
+			obj, err := apiBindingInformer.Informer().GetIndexer().ByIndex(indexers.ByLogicalCluster, clusterName.String())
+			if err != nil {
+				klog.Errorf("Unable to get bindings from indexer: %v", err)
+				return ret, err
+			}
+
+			for _, o := range obj {
+				binding, ok := o.(*apisv1alpha1.APIBinding)
+				if !ok {
+					klog.Errorf("unexpected type apisv1alpha1.APIBinding, got %T", o)
+				}
+				ret = append(ret, binding)
+			}
+
+			return ret, nil
+		},
+		apiBindingsIndexer: apiBindingInformer.Informer().GetIndexer(),
+	}
+
+	apiBindingInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc: func(obj interface{}) { c.enqueueAPIBinding(obj, "") },
+		UpdateFunc: func(_, newObj interface{}) {
+			// TODO: In the future, when we should calculate if there is a need for an update.
+			c.enqueueAPIBinding(newObj, "")
+		},
+		DeleteFunc: func(obj interface{}) { c.enqueueAPIBinding(obj, "") },
+	})
+
+	return c, nil
+
+}
+
+// controller reconciles resource labels that make claimed resources visible to an APIExport
+// owner. It labels resources in the intersection of `APIBinding.status.permissionClaims` and
+// `APIBinding.spec.acceptedPermissionClaims`.
+type controller struct {
+	queue workqueue.RateLimitingInterface
+
+	kcpClusterClient     kcpclient.Interface
+	apiBindingsIndexer   cache.Indexer
+	dynamicClusterClient dynamic.Interface
+	ddsif                *informer.DynamicDiscoverySharedInformerFactory
+
+	apiBindingsLister apislisters.APIBindingLister
+	listAPIBindings   func(clusterName logicalcluster.Name) ([]*apisv1alpha1.APIBinding, error)
+}
+
+// enqueueAPIBinding enqueues an APIBinding .
+func (c *controller) enqueueAPIBinding(obj interface{}, logSuffix string) {
+	key, err := cache.DeletionHandlingMetaNamespaceKeyFunc(obj)
+	if err != nil {
+		runtime.HandleError(err)
+		return
+	}
+
+	klog.V(2).InfoS("queueing APIBinding", "key", key, "suffix", logSuffix)
```

Comment:
nit: I wouldn't call the log key `suffix`. Maybe something like `details`, and only if `logSuffix != ""`.